src/stim/search/sat/wcnf.cc

```diff
@@ -0,0 +1,261 @@
+      clause.add_var(~x);
+      clause.add_var(y);
+      clause.add_var(z);
+      add_clause(clause);
+    }
+    {
+      Clause clause;
+      // hard clause (x, y, z) != (1, 1, 1)
+      clause.add_var(~x);
+      clause.add_var(~y);
+      clause.add_var(~z);
+      add_clause(clause);
+    }
+    return z;
+  }
+
+  size_t quantized_weight(bool weighted, size_t weight_scale_factor, size_t top, Weight weight) {
+    if (weight == HARD_CLAUSE_WEIGHT) {
+      // Hard clause
+      return top;
+    }
+    // Soft clause
+    if (!weighted) {
+      // For unweighted problems, all soft clauses have weight 1.
+      return 1;
+    }
+    return std::round(weight / max_weight * (double)weight_scale_factor);
+  }
+
+  std::string to_wdimacs(bool weighted, size_t weight_scale_factor) {
+    // 'top' is a special weight used to indicate a hard clause.
+    // Should be at least the sum of the weights of all soft clauses plus 1.
+    size_t top;
+    if (weighted) {
+      top = 1 + weight_scale_factor * clauses.size();
+    } else {
+      top = 1 + clauses.size();
+    }
+
+    // WDIMACS header format: p wcnf nbvar nbclauses top
+    // see http://www.maxhs.org/docs/wdimacs.html
+    std::stringstream ss;
+    ss << "p wcnf "<< num_variables << " "<<clauses.size() << " " << top << "\n";
+
+    // Add clauses, 1 on each line.
+    for (const auto& clause : clauses) {
+      // WDIMACS clause format: weight var1 var2 ...
+      // To show negation of a variable, the index should be negated.
+      size_t qw = quantized_weight(weighted, weight_scale_factor, top, clause.weight);
+      // There is no need to add a clause with zero weight.
+      // This can happen if the error has probability 0.5 or if the weight_scale_factor is too
+      // small to accomodate the entire dynamic range of weight values.
+      if (qw == 0) continue;
+      ss << qw;
+      for (size_t i=0; i<clause.vars.size(); ++i) {
+        BoolRef var = clause.vars[i];
+        // Variables are 1-indexed
+        if (var.negated) {
+          ss << " -" << (var.variable + 1);
+        } else {
+          ss << " " << (var.variable + 1);
+        }
+      }
+      // Each clause ends with 0
+      ss << " 0\n";
+    }
+    return ss.str();
+  }
+};
+
+std::string stim::shortest_error_problem_as_wcnf_file(
+  const DetectorErrorModel &model, bool weighted, size_t weight_scale_factor) {
+  MaxSATInstance inst;
+
+  if (weighted and weight_scale_factor < 1) {
+    throw std::invalid_argument("for weighted problems, weight_scale_factor must be >= 1");
+  }
+  if (!weighted and weight_scale_factor != 0) {
+    throw std::invalid_argument("for unweighted problems, weight_scale_factor must be == 0");
+  }
+
+  size_t num_observables = model.count_observables();
+  size_t num_detectors = model.count_detectors();
+  size_t num_errors = model.count_errors();
+  if (num_observables == 0 or num_detectors == 0 or num_errors == 0) {
+    return UNSAT_WCNF_STR;
```

Comment:
This is wrong. The no-det yes-err yes-obs case is often solvable with a result of d=1.

Add unit tests including a circuit with obs but no dets, dets but no obs, no obs and no dets, no errors, etc. Hit the corner cases.

Ideally, the method should not really check for the no-obs case, and instead just do what it normally does resulting in an unsatisfiable problem.

Reply:
> This is wrong. The no-det yes-err yes-obs case is often solvable with a result of d=1.

good catch, fixed!

> Add unit tests including a circuit with obs but no dets, dets but no obs, no obs and no dets, no errors, etc. Hit the corner cases.

Done

> Ideally, the method should not really check for the no-obs case, and instead just do what it normally does resulting in an unsatisfiable problem.

I like your suggestion to 'do what it normally does'. Unfortunately SAT formulas do not permit clauses that involve constants which makes translating this edge case a bit unnatural to handle later in the translation.

If you would like aspects of the implementation to be changed I suggest we defer that to a future PR once the API and tests are solidified here.